### PR TITLE
Add search to getPubs query

### PR DESCRIPTION
**Pull request by allisonking**
## Issue(s) Resolved
`getPubs` portion of https://github.com/pubpub/platform/issues/712

### CODE REVIEW

File: core/lib/server/pub.ts [hunk 1]
```diff
@@ -342,6 +344,9 @@ export const getPubs = async (
 					.where("PubsInStages.stageId", "=", props.stageId!)
 			)
 			.$if(Boolean(params.onlyParents), (eb) => eb.where("pubs.parentId", "is", null))
+			// .$if(Boolean(search), (eb) =>
+			// 	eb.where(sql`pubs.values::jsonb::text`, "ilike", `%${search}%`)
+			// )
```
Comment (allisonking) [inline core/lib/server/pub.ts:349]: this doesn't work—I think I'd like to query `latest_values` but I'm not sure that's exposed at this point. I'm also not sure if this search is quite what we want, or if we only want to be searching on the title of a pub (in which case we wouldn't be able to do this until we have that `isTitle` field). my original idea here was to be able to search all pub values as text just on if the search value was in there at all which might not be performant. but either way, my sql isn't good enough 🥲

Comment (allisonking) [reply core/lib/server/pub.ts:349]: the context editor storybook only filters on title https://github.com/pubpub/platform/blob/3aa56fccd92ebd1c532ea07042b82d0c507d1d92/packages/context-editor/src/stories/mockUtils.ts#L5-L12

File: core/lib/__tests__/pub.test.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+				id: submissionPubId2,
+				pubType: "Submission",
+				values: {
+					Title: "Dogs and cats",
+					Content: "They both like to eat.",
+				},
+			},
+		],
+	});
+	expect(seed).toBeDefined();
```
[the remark this answers is not in the document]
Comment (tefkah) [reply core/lib/__tests__/pub.test.ts:57]: okay this stumped me for a while, but there are two ways to fix this:

1. Move the `seedCommunity` call outside of `test(()=>{})`. This fixes it, but then doesn't clean up the community.
2. Make `getPubFields` accept a `trx`, like so
e.g. by replacing this
https://github.com/pubpub/platform/blob/bd08a370f26b47b434ac5cc3fe7adb815b9d68e0/core/lib/server/pubFields.ts#L19-L68

by

```ts
export const getPubFields = (
	props:
		| {
				pubId?: never;
				pubTypeId?: never;
				communityId?: never;
				includeRelations?: never;
				trx?: Kysely<Database>;
		  }
		| {
				pubId: PubsId;
				valuesOnly?: boolean;
				pubTypeId?: never;
				communityId?: never;
				includeRelations?: never;
				trx?: Kysely<Database>;
		  }
		| {
				pubId?: never;
				pubTypeId: PubTypesId;
				communityId?: never;
				includeRelations?: never;
				trx?: Kysely<Database>;
		  }
		| {
				pubId?: never;
				pubTypeId?: never;
				communityId: CommunitiesId;
				includeRelations?: never;
				trx?: Kysely<Database>;
		  }
		| {
				pubId?: never;
				pubTypeId?: never;
				communityId: CommunitiesId;
				includeRelations: boolean;
				trx?: Kysely<Database>;
		  } = {
		trx: db,
	}
) => autoCache(_getPubFields(props));

export const _getPubFields = (
	props:
		| {
				pubId?: never;
				pubTypeId?: never;
				communityId?: never;
				includeRelations?: never;
				trx?: Kysely<Database>;
		  }
		| {
				pubId: PubsId;
				valuesOnly?: boolean;
				pubTypeId?: never;
				communityId?: never;
				includeRelations?: never;
				trx?: Kysely<Database>;
		  }
		| {
				pubId?: never;
				pubTypeId: PubTypesId;
				communityId?: never;
				includeRelations?: never;
				trx?: Kysely<Database>;
		  }
		| {
				pubId?: never;
				pubTypeId?: never;
				communityId: CommunitiesId;
				includeRelations?: never;
				trx?: Kysely<Database>;
		  }
		| {
				pubId?: never;
				pubTypeId?: never;
				communityId: CommunitiesId;
				includeRelations: boolean;
				trx?: Kysely<Database>;
		  } = { trx: db }
) =>
	(props.trx ?? db)
``` 

and then passing that in here:
 https://github.com/pubpub/platform/blob/5eae87f495d455924b22282d23c5629166ece298/core/lib/server/pub.ts#L413-L416

like so

```ts
	const pubFieldsForCommunityObject = await getPubFields({
		communityId,
		includeRelations: true,
		trx,
	}).executeTakeFirst();
```

[the remark this answers is not in the document]
Comment (tefkah) [reply core/lib/__tests__/pub.test.ts:57]: I'm a bit confused as to why this works though.

I think it's roughly that the `getPubFields` function ends up using a different `db` than the one in `seedCommunity`/`createPubRecursiveNew`. But I don't really know why, I feel like they both should be mocked correctly.

Moving the `seedCommunity` outside of the `test` call makes both of them use the "normal" `db` instance I think, while passing the `trx` makes them use the current transaction.

Hmm but i'm not really sure about all this, i fear we've created a rather complex system 😅

[the remark this answers is not in the document]
Comment (tefkah) [reply core/lib/__tests__/pub.test.ts:57]: okay the craziest way to fix this is by dynamically importing `getPubs` as well

so add
```ts
		const { getPubs } = await import("../server/pub");
```
before the first `getPubs` call, and remove the import.

For some reason this also gets rid of the `No Pubfields found` error?? I really don't know why, my guess is that if even one function that uses `db` is not dynamically imported, then the mock does not succeed reliably.

I would do this as a fix!

**Comment by allisonking** [comment]
thanks for all the help @tefkah !! as discussed today, it's probably too early to work on this bit right now, so I'm going to close this PR. and there's a ticket to do this work here https://github.com/pubpub/platform/issues/737. for the purposes of https://github.com/pubpub/platform/issues/712, we'll keep it stubbed out until 737

**Pull request closed without merging**